cis194/week2/daniel-deng/src/CIS194/Homework02/Exercise01.hs

```diff
@@ -0,0 +1,48 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module CIS194.Homework02.Exercise01 (parseMessage, parseMessages) where
+
+import qualified CIS194.Homework02.Log as Log
+
+import           Data.Text             hiding (lines, map)
+import           Text.Parsec
+import           Text.Parsec.Text
+
+skipSpaces :: Parser ()
+skipSpaces = skipMany1 $ char ' '
+
+skipSpacesIfAny :: Parser ()
+skipSpacesIfAny = skipMany $ char ' '
+
+digitsAfter :: Char -> Parser String
+digitsAfter c = char c *> skipSpaces *> many1 digit
+
+(~>) :: Char -> value -> Parser value
+c ~> t = char c >> pure t
+
+msgTypeParser :: Parser Log.MessageType
+msgTypeParser =
+      'I' ~> Log.Info
+  <|> 'W' ~> Log.Warning
+  <|> Log.Error . read <$> digitsAfter 'E'
+
+tsParser :: Parser Log.TimeStamp
+tsParser = read <$> many1 digit
+
+bodyParser :: Parser String
+bodyParser = manyTill anyChar eof
+
+logMsgParser :: Parser Log.LogMessage
+logMsgParser =
+  Log.LogMessage
+  <$> msgTypeParser <* skipSpaces
+  <*> tsParser <* skipSpacesIfAny
+  <*> bodyParser
+
+parseMessage :: String -> Log.LogMessage
+parseMessage msgLine = case parse logMsgParser "" (pack msgLine) of
+                         Right logMsg -> logMsg
+                         _            -> Log.Unknown msgLine
+
+parseMessages :: String -> [Log.LogMessage]
+parseMessages = map parseMessage . lines
```

Comment:
We prefer polymorphic than monomorphic function most the time
```suggestion
parseMessages = fmap parseMessage . lines
```

Reply:
@shineli1984 so for all previous homework we should've used `fmap` instead of the binding `map` for list operations?

Reply:
Yes. `fmap` is preferred in most cases.